File: ylesanne1/Geom2.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ylesanne1_2
{
    class Point2
    {
        public double x { get; set; }
        public double y { get; set; }
        public double _rho { get; set; }
        public double _theta { get; set; }

        public Point2(double x = 0.0, double y = 0.0)
        {
            _rho = Math.Sqrt(Math.Pow(x, 2) + Math.Pow(y, 2));
            _theta = Math.Atan2(y, x);


        }
        //queries
        public override string ToString()
        {
            return string.Format("x: {0}\n" + "y: {1}\n" + "rho: {2}\n" + "theta: {3}", getX(), getY(), rho(), theta());
        }

        public double getX()
        {
            return _rho * Math.Cos(_theta);
        }

        public double getY()
        {
            return rho() * Math.Sin(theta());
        }
       
        public double rho()
        {
            return _rho;
        }

        public double theta()
        {
            return _theta;
        }

        public double distance(Point2 p)
        {
            return vectorTo(p).rho();
        }

        public Point2 vectorTo(Point2 p)
        {
            return new Point2(p.getX() - getX(), p.getY() - getY());
        }

        //commands
        public void translate(double dx, double dy)
        {
            x = getX() + dx;
            y = getY() + dy;
            _rho = Math.Sqrt(Math.Pow(x, 2) + Math.Pow(y, 2));
            _theta = Math.Atan2(y, x);

        }

        public void scale(double factor)
        {
            _rho *= factor;
        }

        public void centre_rotate(double angle)
        {
            _theta = (_theta + angle) % (2 * Math.PI);
        }

        public void rotate(Point2 p, double angle)
        {
            translate(-p.getX(), -p.getY());
            centre_rotate(angle);
            translate(p.getX(), p.getY());
        }
    }
}

File: ylesanne1/Geom2.py
import math

class Point:
    "Two-dimensional points"

    def __init__(self, x=0.0, y=0.0):
        self._rho = math.sqrt(x**2 + y**2)
        self._theta = math.atan2(y, x)

    def __str__(self):
        result = "\n".join(["x: %f" % self.x(),
                            "y: %f" % self.y(),
                            "rho: %f" % self.rho(),
                            "theta: %f" % self.theta()])
        return result

# Queries

    def x(self):
        "Abscissa"
        return self.rho()*math.cos(self.theta())

    def y(self):
        "Ordinate"
        return self.rho()*math.sin(self.theta())

    def rho(self):
        "Distance to origin (0, 0)"
        return self._rho

    def theta(self):
        "Angle to horizontal axis"
        return self._theta

    #Precondition: 	    other != none
	#Postcondition:     vectorTo(other).rho()
    def distance(self, other):
        "Distance to other"
        return self.vectorTo(other).rho()

    #Precondition: 	    other != none
	#Postcondition:     x() = other.x() - x()
	#                   y() = other.y() - y()
    def vectorTo(self, other):
        "Returns the Point representing the vector from self to other Point"
        return Point(other.x() - self.x(), other.y() - self.y())

# Commands

	#Precondition: 	    -
	#Postcondition:	    x() = old x() + dx
	#		            y() = old y() + dy 
    def translate(self, dx, dy):
        "Move by dx horizontally, dy vertically"
        x = self.x() + dx
        y = self.y() + dy
        self._rho = math.sqrt(x**2 + y**2)
        self._theta = math.atan2(y, x)

    #Precondition: 	    -
	#Postcondition:	    rho = old Rho() * factor
    def scale(self, factor):
        "Scale by factor"
        self._rho *= factor

    #Precondition:	    angle != none
	#Postcondition:	    theta() = (old theta() + angle) % (2*math.pi)
    def centre_rotate(self, angle):
        "Rotate around origin (0, 0) by angle"
        self._theta = (self._theta + angle) % (2 * math.pi)

    #Precondition:	    p != none
	#		            angle != none
    #Postcondition:	    p.vectorTo(this).theta() = (p.vectorTo(old this).theta() + angle) % (2*math.pi)
    def rotate(self, p, angle):
        "Rotate around p by angle"
        self.translate(-p.x(), -p.y())
        self.centre_rotate(angle)
        self.translate(p.x(), p.y())


#"Invariandid (tingimused, mis kehtivad kogu aeg):"
#x() = rho() * cos(theta())
#y() = rho() * sin(theta())
#rho() = sqrt(x() * x() + y() * y())
#theta() = atan2(y(),x())

